contracts/src/TeleporterTokenSource.sol

```diff
@@ -182,9 +305,55 @@ abstract contract TeleporterTokenSource is
         );
 
         if (isMultihop) {
-            emit TokensAndCallRouted(messageID, input, amount);
+            emit TokensAndCallRouted(messageID, input, adjustedAmount);
         } else {
-            emit TokensAndCallSent(messageID, originSenderAddress, input, amount);
+            emit TokensAndCallSent(messageID, originSenderAddress, input, adjustedAmount);
+        }
+    }
+
+    /**
+     * @dev See {INativeTokenSource-addCollateral}
+     */
+    function _addCollateral(
+        bytes32 destinationBlockchainID,
+        address destinationBridgeAddress,
+        uint256 amount
+    ) internal sendNonReentrant {
+        DestinationBridgeSettings memory destinationSettings =
+            registeredDestinations[destinationBlockchainID][destinationBridgeAddress];
+        require(
+            destinationSettings.registered,
+            "TeleporterTokenSource: destination bridge not registered"
+        );
+        require(
+            destinationSettings.collateralNeeded > 0,
+            "TeleporterTokenSource: zero collateral needed"
+        );
+
+        // Deposit the full amount, and withdraw back to the sender if there is excess.
+        amount = _deposit(amount);
+
+        // Calculate the amount remaining.
+        uint256 remainingCollateralNeeded;
+        uint256 excessAmount;
+        if (amount >= destinationSettings.collateralNeeded) {
+            remainingCollateralNeeded = 0;
+            excessAmount = amount - destinationSettings.collateralNeeded;
+            amount = destinationSettings.collateralNeeded;
```

Comment:
Can we leave a comment that `amount` is being used for how much collateral has been added?